README
-*- coding: utf-8-unix -*-

			       CCLanguage
			       ==========


Topics
------

  1. Introduction
  2. License
  3. Install
  4. Usage
  A. Credits
  B. Bugs, vulnerabilities and contributions
  C. Resources
  D. Badges and static analysis


1. Introduction
---------------

This  package  installs a  C11  language  library that  implements  some
adaptations and extensions to the language.

  The package targets POSIX systems.  The package uses the GNU Autotools
and it is tested, using Travis  CI, on both Ubuntu GNU+Linux systems and
OS X systems.


2. License
----------

Copyright (c) 2020 Marco Maggi <mrc.mgg@gmail.com>

  This program is  free software: you can redistribute  it and/or modify
it under the terms of the GNU Lesser General Public License as published
by the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

  This program  is distributed in the  hope that it will  be useful, but
WITHOUT   ANY   WARRANTY;  without   even   the   implied  warranty   of
MERCHANTABILITY  or  FITNESS FOR  A  PARTICULAR  PURPOSE.  See  the  GNU
General Public License for more details.

  You  should have  received a  copy of  the GNU  Lesser General  Public
License     along     with     this    program.      If     not,     see
<http://www.gnu.org/licenses/>.


3. Install
----------

To install from a proper release tarball, do this:

    $ cd cclanguage-0.1.0
    $ mkdir build
    $ cd build
    $ ../configure
    $ make
    $ make check
    $ make install

to inspect the available configuration options:

    $ ../configure --help

  The Makefile is designed to allow parallel builds, so we can do:

   $ make -j4 all && make -j4 check

which,  on  a  4-core  CPU,   should  speed  up  building  and  checking
significantly.

  The  Makefile supports  the  DESTDIR environment  variable to  install
files in a temporary location, example: to see what will happen:

    $ make -n install DESTDIR=/tmp/cclanguage

to really do it:

    $ make install DESTDIR=/tmp/cclanguage

  After the installation it is  possible to verify the installed library
against the test suite with:

    $ make installcheck

  From  a repository  checkout or  snapshot  (the ones  from the  GitHub
site): we must  install the GNU Autotools (GNU  Automake, GNU Autoconf),
then  we must  first run  the script  "autogen.sh" from  the top  source
directory, to generate the needed files:

    $ cd cclanguage
    $ sh autogen.sh

notice  that   "autogen.sh"  will  run  the   program  "autoreconf"  and
"libtoolize"; the  latter is  selected through the  environment variable
"LIBTOOLIZE",  whose  value  can  be  customised;  for  example  to  run
"glibtoolize" rather than "libtoolize" we do:

   $ LIBTOOLIZE=glibtoolize sh autogen.sh

  After this the  procedure is the same  as the one for  building from a
proper release tarball, but we have to enable maintainer mode:

    $ ../configure --enable-maintainer-mode [options]
    $ make
    $ make check
    $ make install


4. Usage
--------

Read the documentation generated from  the Texinfo sources.  The package
installs the documentation  in Info format; we can  generate and install
documentation in HTML format by running:

   $ make html
   $ make install-html


A. Credits
----------

The  stuff was  written by  Marco Maggi.   If this  package exists  it's
because of the great GNU software tools that he uses all the time.


B. Bugs
-------

Bug  and vulnerability  reports are  appreciated, all  the vulnerability
reports  are  public; register  them  using  the  Issue Tracker  at  the
project's GitHub  site.  For  contributions and  patches please  use the
Pull Requests feature at the project's GitHub site.


C. Resources
------------

The latest release of this package can be downloaded from:

	<https://bitbucket.org/marcomaggi/cclanguage/downloads>

development takes place at:

	       <http://github.com/marcomaggi/cclanguage/>

and as backup at:

	     <https://bitbucket.org/marcomaggi/cclanguage/>

the documentation is available online:

	   <http://marcomaggi.github.io/docs/cclanguage.html>

the GNU Project software can be found here:

			 <http://www.gnu.org/>


D. Badges and static analysis
-----------------------------

D.1 Travis CI
-------------

Travis CI is  a hosted, distributed continuous  integration service used
to build and test software projects  hosted at GitHub.  We can find this
project's dashboard at:

	    <https://travis-ci.org/marcomaggi/cclanguage>

  Usage of this service is configured through the file ".travis.yml" and
the scripts under the directory "meta/travis-ci".


D.2 Clang's Static Analyzer
---------------------------

The Clang Static Analyzer is a source code analysis tool that finds bugs
in C, C++, and Objective-C programs.  It is distributed along with Clang
and we can find it at:

		   <http://clang-analyzer.llvm.org/>

  Usage of this service is implemented with make rules; see the relevant
section in the file "Makefile.am".


D.3 Codecov
-----------

Codecov is a service providing code  coverage reports.  We can find this
project's dashboard at:

	     <https://codecov.io/gh/marcomaggi/cclanguage>

Usage of  this service is  implemented through direct  interface between
GitHub and Codecov  sites; it configured through  the file "codecov.yml"
and appropriate entries in Travis CI's matrix of builds.


D.4 Codacy
----------

Codacy is  an online service  providing code  review.  We can  find this
project's dashboard at:

	   <https://www.codacy.com/app/marcomaggi/cclanguage>

Usage of this service is  implemented through direct integration between
GitHub and Codacy sites.


### end of file
# Local Variables:
# mode: text
# fill-column: 72
# End:


NEWS
-*- coding: utf-8-unix -*-


			News file for CCLanguage
			========================


Notes for revision 0.1.0-devel.0
--------------------------------

* Initial release.  Most  of the code and documentation  in this package
  were distributed in the package  CCExceptions before.  Now they are in
  a standalone package.


### end of file
# Local Variables:
# mode: text
# fill-column: 72
# paragraph-start: "*"
# ispell-local-dictionary: "en_GB"
# End:


tests/version.c
/*
  Part of: CCLanguage
  Contents: test for version functions
  Date: Jun  9, 2020

  Abstract

	Test file for version functions.

  Copyright (C) 2020 Marco Maggi <mrc.mgg@gmail.com>

  See the COPYING file.
*/

#include <cclanguage.h>
#include <stdio.h>
#include <stdlib.h>

int
main (int argc CCLIB_UNUSED, char const *const argv[] CCLIB_UNUSED)
{
  printf("version number string: %s\n", cclang_version_string());
  printf("libtool version number: %d:%d:%d\n",
	 cclang_version_interface_current(),
	 cclang_version_interface_revision(),
	 cclang_version_interface_age());
  exit(EXIT_SUCCESS);
}

/* end of file */


tests/destructor-trait.h
/*
  Part of: CCLanguage
  Contents: example definition of a destructor trait
  Date: Apr 12, 2020

  Abstract

	This  header  file  defines  the  trait  "my_destructor_T".   It  is  a  data
	structure, passed  as value, using the  common structs API, using  the struct
	descriptor API.   The purpose of  this trait is to  provide a unified  API to
	call the destructor of any resource.

  Copyright (C) 2020 Marco Maggi <mrc.mgg@gmail.com>

  The author  hereby grant permission to  use, copy, modify, distribute,  and license
  this  software  and its  documentation  for  any  purpose, provided  that  existing
  copyright  notices are  retained in  all copies  and that  this notice  is included
  verbatim in  any distributions. No  written agreement,  license, or royalty  fee is
  required for  any of the  authorized uses.  Modifications  to this software  may be
  copyrighted by  their authors  and need  not follow  the licensing  terms described
  here, provided that the  new terms are clearly indicated on the  first page of each
  file where they apply.

  IN NO  EVENT SHALL THE AUTHOR  OR DISTRIBUTORS BE  LIABLE TO ANY PARTY  FOR DIRECT,
  INDIRECT, SPECIAL, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING OUT OF  THE USE OF
  THIS SOFTWARE,  ITS DOCUMENTATION, OR ANY  DERIVATIVES THEREOF, EVEN IF  THE AUTHOR
  HAVE BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

  THE AUTHOR  AND DISTRIBUTORS SPECIFICALLY  DISCLAIM ANY WARRANTIES,  INCLUDING, BUT
  NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
  PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE IS  PROVIDED ON AN "AS IS" BASIS, AND
  THE AUTHOR  AND DISTRIBUTORS  HAVE NO OBLIGATION  TO PROVIDE  MAINTENANCE, SUPPORT,
  UPDATES, ENHANCEMENTS, OR MODIFICATIONS.

*/

#ifndef DESTRUCTOR_TRAIT_H
#define DESTRUCTOR_TRAIT_H

#ifdef __cplusplus
extern "C" {
#endif


/** --------------------------------------------------------------------
 ** Descrs.
 ** ----------------------------------------------------------------- */

#include <cclanguage.h>
#include <stdio.h>
#include <stdlib.h>


/** --------------------------------------------------------------------
 ** Destructor trait.
 ** ----------------------------------------------------------------- */

CCLIB_DEFINE_TRAIT(my_destructor_T);

/* A typedef for every method. */
typedef void                            cclib_method_type(my_destructor_T, destroy) (my_destructor_T self);

/* The methods table declaration. */
struct cclib_methods_table_type(my_destructor_T) {
  cclib_method_type(my_destructor_T, destroy) *	destroy;
};


/** --------------------------------------------------------------------
 ** Done.
 ** ----------------------------------------------------------------- */

#ifdef __cplusplus
} // extern "C"
#endif

#endif /* define DESTRUCTOR_TRAIT_H */

/* end of file */


src/cclanguage.h
/*
  Part of: CCLanguage
  Contents: public header file for CCLanguage projects
  Date: Mar 30, 2020

  Abstract

	This header file contains  definitions to be used by all  the packages in the
	CCLibraries project.

  Copyright (C) 2020 Marco Maggi <mrc.mgg@gmail.com>

  This is free software; you can redistribute  it and/or modify it under the terms of
  the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3.0 of the License, or (at your option) any later version.

  This library  is distributed in the  hope that it  will be useful, but  WITHOUT ANY
  WARRANTY; without  even the implied  warranty of  MERCHANTABILITY or FITNESS  FOR A
  PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License along with
  this library; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
  Suite 330, Boston, MA 02111-1307 USA.
*/

#ifndef CCLANGUAGE_H
#define CCLANGUAGE_H 1

#ifdef __cplusplus
extern "C" {
#endif


/** --------------------------------------------------------------------
 ** Helper macros.
 ** ----------------------------------------------------------------- */

/* Pointer cast macro helper. */
#define CCLIB_PC(TYPE,X,Y)		TYPE * X = (TYPE *) (Y)

/* Type cast macro helper. */
#define CCLIB_CAST(TYPE,X,Y)		TYPE X = (TYPE) (Y)


/** --------------------------------------------------------------------
 ** Header files.
 ** ----------------------------------------------------------------- */

#include <cclanguage-compiler.h>
#include <cclanguage-typedefs.h>
#include <cclanguage-variadic-macros.h>
#include <cclanguage-structs.h>


/** --------------------------------------------------------------------
 ** Version functions.
 ** ----------------------------------------------------------------- */

cclib_decl char const *	cclang_version_string			(void);
cclib_decl int		cclang_version_interface_current	(void);
cclib_decl int		cclang_version_interface_revision	(void);
cclib_decl int		cclang_version_interface_age		(void);


/** --------------------------------------------------------------------
 ** Done.
 ** ----------------------------------------------------------------- */

#ifdef __cplusplus
} // extern "C"
#endif

#endif /* defined CCLANGUAGE_H */

/* end of file */


src/cclanguage-library.c
/*
  Part of: CCLanguage
  Contents: library functions
  Date: Jun  9, 2020

  Abstract

	This module implements library initialisation and version numbers inspection.

  Copyright (C) 2020 Marco Maggi <mrc.mgg@gmail.com>

  This program is free  software: you can redistribute it and/or  modify it under the
  terms of the  GNU Lesser General Public  License as published by  the Free Software
  Foundation, either version 3 of the License, or (at your option) any later version.

  This program  is distributed in the  hope that it  will be useful, but  WITHOUT ANY
  WARRANTY; without  even the implied  warranty of  MERCHANTABILITY or FITNESS  FOR A
  PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License along with
  this program.  If not, see <http://www.gnu.org/licenses/>.
*/


/** --------------------------------------------------------------------
 ** Headers.
 ** ----------------------------------------------------------------- */

#include "cclanguage-internals.h"


/** --------------------------------------------------------------------
 ** Version functions.
 ** ----------------------------------------------------------------- */

char const *
cclang_version_string (void)
{
  return cclanguage_VERSION_INTERFACE_STRING;
}
int
cclang_version_interface_current (void)
{
  return cclanguage_VERSION_INTERFACE_CURRENT;
}
int
cclang_version_interface_revision (void)
{
  return cclanguage_VERSION_INTERFACE_REVISION;
}
int
cclang_version_interface_age (void)
{
  return cclanguage_VERSION_INTERFACE_AGE;
}


/** --------------------------------------------------------------------
 ** Library initialisation.
 ** ----------------------------------------------------------------- */

/*
void
cclang_library_init (void)
{
  static bool	to_be_initialised = true;

  if (to_be_initialised) {
    to_be_initialised = false;
    ccstr_library_init();
  }
}

*/

/* end of file */


src/cclanguage-typedefs.h
/*
  Part of: CCLanguage
  Contents: public header file for CCLibraries projects
  Date: Jun  5, 2020

  Abstract

	This header file contains  definitions to be used by all  the packages in the
	CCLibraries project.  It defines some common types.

  Copyright (C) 2020 Marco Maggi <mrc.mgg@gmail.com>

  This is free software; you can redistribute  it and/or modify it under the terms of
  the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3.0 of the License, or (at your option) any later version.

  This library  is distributed in the  hope that it  will be useful, but  WITHOUT ANY
  WARRANTY; without  even the implied  warranty of  MERCHANTABILITY or FITNESS  FOR A
  PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License along with
  this library; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
  Suite 330, Boston, MA 02111-1307 USA.
*/

#ifndef CCLANGUAGE_TYPEDEFS_H
#define CCLANGUAGE_TYPEDEFS_H 1

#ifdef __cplusplus
extern "C" {
#endif


/** --------------------------------------------------------------------
 ** Header files.
 ** ----------------------------------------------------------------- */

#include <cclanguage-platform.h>


/** --------------------------------------------------------------------
 ** Helper macros.
 ** ----------------------------------------------------------------- */

#define CCLIB_STRUCT_TYPEDEF(STRUCT)	typedef struct STRUCT STRUCT


/** --------------------------------------------------------------------
 ** Common type definitions: opaque data structures handling.
 ** ----------------------------------------------------------------- */

typedef void				cclib_resource_data_t;
typedef cclib_resource_data_t *		cclib_resource_pointer_t;

#undef  cclib_resource_pointer
#define cclib_resource_pointer(EXPR)	((cclib_resource_pointer_t)(EXPR))


/** --------------------------------------------------------------------
 ** Done.
 ** ----------------------------------------------------------------- */

#ifdef __cplusplus
} // extern "C"
#endif

#endif /* defined CCLANGUAGE_TYPEDEFS_H */

/* end of file */


src/cclanguage-structs.h
/*
  Part of: CCLanguage
  Contents: public header file for CCLibraries projects
  Date: Jun  5, 2020

  Abstract

	This header file contains  definitions to be used by all  the packages in the
	CCLibraries  project.  This  header files  defines varidic  macros for  names
	generation related to common data structures APIs.

  Copyright (C) 2020 Marco Maggi <mrc.mgg@gmail.com>

  This is free software; you can redistribute  it and/or modify it under the terms of
  the GNU Lesser General Public License as published by the Free Software Foundation;
  either version 3.0 of the License, or (at your option) any later version.

  This library  is distributed in the  hope that it  will be useful, but  WITHOUT ANY
  WARRANTY; without  even the implied  warranty of  MERCHANTABILITY or FITNESS  FOR A
  PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License along with
  this library; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
  Suite 330, Boston, MA 02111-1307 USA.
*/

#ifndef CCLANGUAGE_STRUCTS_H
#define CCLANGUAGE_STRUCTS_H 1

#ifdef __cplusplus
extern "C" {
#endif


/** --------------------------------------------------------------------
 ** Automatically generated names API: data structs.
 ** ----------------------------------------------------------------- */

/* Whether it appears to make sense or not, let's always implement 4 variants of each
   "well  known" macro  name: without  VAR, with  single VAR,  with double  VAR, with
   triple VAR. */

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "make()". */
#define cclib_make_1(STRUCT)				STRUCT ## __make
#define cclib_make_2(STRUCT, VAR)			STRUCT ## __make__ ## VAR
#define cclib_make_3(STRUCT, VAR1, VAR2)		STRUCT ## __make__ ## VAR1 ## _ ## VAR2
#define cclib_make_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __make__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_make(...)					CCLIB_VNAME(cclib_make, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "unmake()". */
#define cclib_unmake_1(STRUCT)				STRUCT ## __unmake
#define cclib_unmake_2(STRUCT, VAR)			STRUCT ## __unmake__ ## VAR
#define cclib_unmake_3(STRUCT, VAR1, VAR2)		STRUCT ## __unmake__ ## VAR1 ## _ ## VAR2
#define cclib_unmake_4(STRUCT, VAR1, VAR2, VAR3)	STRUCT ## __unmake__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_unmake(...)				CCLIB_VNAME(cclib_unmake, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "new()". */
#define cclib_new_1(STRUCT)				STRUCT ## __new
#define cclib_new_2(STRUCT, VAR)			STRUCT ## __new__ ## VAR
#define cclib_new_3(STRUCT, VAR1, VAR2)			STRUCT ## __new__ ## VAR1 ## _ ## VAR2
#define cclib_new_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __new__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_new(...)					CCLIB_VNAME(cclib_new, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "delete()". */
#define cclib_delete_1(STRUCT)				STRUCT ## __delete
#define cclib_delete_2(STRUCT, VAR)			STRUCT ## __delete__ ## VAR
#define cclib_delete_3(STRUCT, VAR1, VAR2)		STRUCT ## __delete__ ## VAR1 ## _ ## VAR2
#define cclib_delete_4(STRUCT, VAR1, VAR2, VAR3)	STRUCT ## __delete__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_delete(...)				CCLIB_VNAME(cclib_delete, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "init()". */
#define cclib_init_1(STRUCT)				STRUCT ## __init
#define cclib_init_2(STRUCT, VAR)			STRUCT ## __init__ ## VAR
#define cclib_init_3(STRUCT, VAR1, VAR2)		STRUCT ## __init__ ## VAR1 ## _ ## VAR2
#define cclib_init_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __init__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_init(...)					CCLIB_VNAME(cclib_init, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "final()". */
#define cclib_final_1(STRUCT)				STRUCT ## __final
#define cclib_final_2(STRUCT, VAR)			STRUCT ## __final__ ## VAR
#define cclib_final_3(STRUCT, VAR1, VAR2)		STRUCT ## __final__ ## VAR1 ## _ ## VAR2
#define cclib_final_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __final__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_final(...)				CCLIB_VNAME(cclib_final, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "alloc()". */
#define cclib_alloc_1(STRUCT)				STRUCT ## __alloc
#define cclib_alloc_2(STRUCT, VAR)			STRUCT ## __alloc__ ## VAR
#define cclib_alloc_3(STRUCT, VAR1, VAR2)		STRUCT ## __alloc__ ## VAR1 ## _ ## VAR2
#define cclib_alloc_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __alloc__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_alloc(...)				CCLIB_VNAME(cclib_alloc, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the API function "release()". */
#define cclib_release_1(STRUCT)				STRUCT ## __release
#define cclib_release_2(STRUCT, VAR)			STRUCT ## __release__ ## VAR
#define cclib_release_3(STRUCT, VAR1, VAR2)		STRUCT ## __release__ ## VAR1 ## _ ## VAR2
#define cclib_release_4(STRUCT, VAR1, VAR2, VAR3)	STRUCT ## __release__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_release(...)				CCLIB_VNAME(cclib_release, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into a function name. */
#define cclib_fun_1(STRUCT)				STRUCT ## __fun
#define cclib_fun_2(STRUCT, VAR)			STRUCT ## __fun__ ## VAR
#define cclib_fun_3(STRUCT, VAR1, VAR2)			STRUCT ## __fun__ ## VAR1 ## _ ## VAR2
#define cclib_fun_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __fun__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_fun(...)					CCLIB_VNAME(cclib_fun, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into a function name to be used for the type predicates. */
#define cclib_is_1(STRUCT)				STRUCT ## __is
#define cclib_is_2(STRUCT, VAR)				STRUCT ## __is__ ## VAR
#define cclib_is_3(STRUCT, VAR1, VAR2)			STRUCT ## __is__ ## VAR1 ## _ ## VAR2
#define cclib_is_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __is__ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_is(...)					CCLIB_VNAME(cclib_is, __VA_ARGS__)


/** --------------------------------------------------------------------
 ** Automatically generated names API: data structure with descriptor.
 ** ----------------------------------------------------------------- */

/* Given a data  structure name STRUCT, which  is meant to be a  "struct with descr":
   expand into the type name of its descriptor field. */
#define cclib_struct_descriptor_type(STRUCT)			cclib_struct_descriptor_ ## STRUCT ## _descr_t
#define cclib_struct_descriptor_t(STRUCT)			cclib_struct_descriptor_ ## STRUCT ## _descr_t

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the struct's methods table type. */
#define cclib_methods_table_type_1(STRUCT)			STRUCT ## __methods_table_t
#define cclib_methods_table_type_2(STRUCT, VAR)			STRUCT ## __methods_table_t__ ## VAR
#define cclib_methods_table_type_3(STRUCT, VAR1, VAR2)		STRUCT ## __methods_table_t__ ## VAR1 ## _ ## VAR2
#define cclib_methods_table_type_4(STRUCT, VAR1, VAR2, VAR3)	STRUCT ## __methods_table_t__ ## VAR1 ## _ ## VAR2 ## __ ## VAR3
#define cclib_methods_table_type(...)				CCLIB_VNAME(cclib_methods_table_type, __VA_ARGS__)
#define cclib_methods_table_t(...)				CCLIB_VNAME(cclib_methods_table_type, __VA_ARGS__)

/* Given a struct type name STRUCT  and an optional variant specification VAR: expand
   into the name of the struct's methods table. */
#define cclib_methods_table_1(STRUCT)				STRUCT ## __methods_table
#define cclib_methods_table_2(STRUCT, VAR)			STRUCT ## __methods_table__ ## VAR
#define cclib_methods_table_3(STRUCT, VAR1, VAR2)		STRUCT ## __methods_table__ ## VAR1 ## _ ## VAR2
#define cclib_methods_table_4(STRUCT, VAR1, VAR2, VAR3)		STRUCT ## __methods_table__ ## VAR1 ## _ ## VAR2 ## __ ## VAR3
#define cclib_methods_table(...)				CCLIB_VNAME(cclib_methods_table, __VA_ARGS__)

/* Given the  struct type name  STRUCT, the method  name METHOD, an  optional variant
   specification  VAR: expand  into  the type  name  of that  variant  of the  method
   function for the struct type. */
#define cclib_method_type_2(STRUCT, METHOD)			STRUCT ## __method_t__ ## METHOD
#define cclib_method_type_3(STRUCT, METHOD, VAR)		STRUCT ## __method_t__ ## METHOD ## __ ## VAR
#define cclib_method_type_4(STRUCT, METHOD, VAR1, VAR2)		STRUCT ## __method_t__ ## METHOD ## __ ## VAR1 ## _ ## VAR2
#define cclib_method_type_5(STRUCT, METHOD, VAR1, VAR2, VAR3)	STRUCT ## __method_t__ ## METHOD ## __ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_method_type(...)					CCLIB_VNAME(cclib_method_type, __VA_ARGS__)
#define cclib_method_t(...)					CCLIB_VNAME(cclib_method_type, __VA_ARGS__)

/* Given  a struct  type  name STRUCT,  a  method name  METHOD,  an optional  variant
   specification VAR: expand into the name of the method for that type. */
#define cclib_method_2(STRUCT, METHOD)				STRUCT ## __method__ ## METHOD
#define cclib_method_3(STRUCT, METHOD, VAR)			STRUCT ## __method__ ## METHOD ## __ ## VAR
#define cclib_method_4(STRUCT, METHOD, VAR1, VAR2)		STRUCT ## __method__ ## METHOD ## __ ## VAR1 ## _ ## VAR2
#define cclib_method_5(STRUCT, METHOD, VAR1, VAR2, VAR3)	STRUCT ## __method__ ## METHOD ## __ ## VAR1 ## _ ## VAR2 ## _ ## VAR3
#define cclib_method(...)					CCLIB_VNAME(cclib_method, __VA_ARGS__)


/** --------------------------------------------------------------------
 ** Automatically generated definitions API: data structure with descriptor.
 ** ----------------------------------------------------------------- */

/* Define everything needed to declare a "struct with descriptor":
 *
 * - The struct typedef.
 * - The typedef of the methods table.
 * - The typedef of the descriptor struct.
 * - The descriptor struct definition.
 *
 * We can define a "struct with descr" as follows:
 *
 *   CCLIB_DEFINE_STRUCT_WITH_DESCRIPTOR(my_complex_t);
 *
 *   struct my_complex_t {
 *     cclib_struct_descriptor(my_complex_t);
 *     double  re;
 *     double  im;
 *   };
 */
#undef  CCLIB_DEFINE_STRUCT_WITH_DESCRIPTOR
#define CCLIB_DEFINE_STRUCT_WITH_DESCRIPTOR(STRUCT)					\
  CCLIB_STRUCT_TYPEDEF(cclib_methods_table_type(STRUCT));				\
											\
  CCLIB_STRUCT_TYPEDEF(cclib_struct_descriptor_type(STRUCT));				\
											\
  struct cclib_struct_descriptor_type(STRUCT) {						\
    cclib_methods_table_type(STRUCT) const	*cclib_table_of_methods_pointer;	\
  }

/* Given the  name of a data  structure STRUCT, which is  meant to be a  "struct with
   descr": expand into the descriptor struct field declaration. */
#define cclib_struct_descriptor(STRUCT)		cclib_struct_descriptor_type(STRUCT) cclib_struct_descr

/* Given a pointer do data structure PTR, which is meant to be a "struct with descr":
   return a pointer to its methods table usable as both lvalue and rvalue. */
#define cclib_struct_descriptor_ref_methods_table_pointer(PTR)	\
  ((PTR)->cclib_struct_descr.cclib_table_of_methods_pointer)

/* Given a pointer do data structure PTR, which is meant to be a "struct with descr":
   set its pointer to methods table to METHODS_TABLE_POINTER. */
#define cclib_struct_descriptor_set_methods_table_pointer(PTR, METHODS_TABLE_POINTER)	\
  (cclib_struct_descriptor_ref_methods_table_pointer(PTR) = (METHODS_TABLE_POINTER))

/* Given a struct type  name STRUCT, a method name METHOD:  expand into an expression
   evaluating to a pointer to the method implementation. */
#define cclib_method_pointer(PTR, METHOD)	(cclib_struct_descriptor_ref_methods_table_pointer(PTR)->METHOD)

/* Given the name of a variable PTR holding  a data structure pointer and the name of
   a method METHOD: expand into an expression evaluating to the pointer to the method
   implementation in the table of methods, usable to call the method. */
#define cclib_call_worker(METHOD, PTR, ...)	cclib_method_pointer(PTR, METHOD)
#define cclib_call(METHOD, ...)			(cclib_call_worker(METHOD, __VA_ARGS__, CCLIB_DUMMY)(__VA_ARGS__))

/* Given the name of a variable VAR holding a data structure and the name of a method
   METHD:  expand  into  an  expression  evaluating to  the  pointer  to  the  method
   implementation in the table of methods, usable to call the method itself. */
#define cclib_vcall_worker(METHOD, VAR, ...)	cclib_method_pointer(&(VAR), METHOD)
#define cclib_vcall(METHOD, ...)		(cclib_vcall_worker(METHOD, __VA_ARGS__, CCLIB_DUMMY)(__VA_ARGS__))


/** --------------------------------------------------------------------
 ** Helper macros: trait values.
 ** ----------------------------------------------------------------- */

#undef  CCLIB_DEFINE_TRAIT
#define CCLIB_DEFINE_TRAIT(TRAIT)										\
  CCLIB_DEFINE_STRUCT_WITH_DESCRIPTOR(TRAIT);									\
  CCLIB_STRUCT_TYPEDEF(TRAIT);											\
														\
  struct TRAIT {												\
    cclib_struct_descriptor(TRAIT);										\
    cclib_resource_data_t const *self;										\
  };														\
														\
  CCLIB_FUNC_ATTRIBUTE_UNUSED											\
  CCLIB_FUNC_ATTRIBUTE_ALWAYS_INLINE										\
  CCLIB_FUNC_ATTRIBUTE_NONNULL(1,2)										\
    static inline TRAIT												\
    cclib_make(TRAIT) (cclib_resource_data_t const * self, cclib_methods_table_type(TRAIT) const * methods)	\
  {														\
    TRAIT	impl = { .self = self };									\
														\
    cclib_struct_descriptor_set_methods_table_pointer(&impl, methods);						\
    return impl;												\
  }														\
														\
  /* We want the use of  this macro to be followed by a semicolon.   So we put this				\
     harmless declaration at the end. */									\
  struct TRAIT

#define cclib_trait_resource_pointer(TRAIT)		((TRAIT).self)


/** --------------------------------------------------------------------
 ** Done.
 ** ----------------------------------------------------------------- */

#ifdef __cplusplus
} // extern "C"
#endif

#endif /* defined CCLANGUAGE_STRUCTS_H */

/* end of file */
